File: src/scanner.cpp
#include "scanner.hpp"

#include <string>
#include <unordered_map>
#include <vector>

#include "main.hpp"
#include "token.hpp"

void Scanner::addToken(TokenType type) {
    addToken(type, Object());
}

void Scanner::addToken(TokenType type, Object obj) {
    std::string text = source.substr(start, current - start);
    tokens.push_back({type, line, text, obj});
}

bool Scanner::isAtEnd() {
    return current >= source.length();
}

bool Scanner::isDigit(char c) {
   return c >= '0' && c <= '9'; 
}

bool Scanner::isAlpha(char c) {
    return (c >= 'a' && c <= 'z') ||
           (c >= 'A' && c <= 'Z') ||
            c == '_';
}

bool Scanner::isAlphaNumeric(char c) {
    return isAlpha(c) || isDigit(c);
}

char Scanner::advance() {
    return source.at(current++);
}

char Scanner::peek() {
    if (isAtEnd()) return '\0';
    return source.at(current);
}

char Scanner::peek(int count) {
    if (current + count >= source.length()) return '\0';
    return source.at(current + count);
}

bool Scanner::match(char expected) {
    if (isAtEnd()) return false;
    if (source.at(current) != expected) return false;

    current++;
    return true;
}


void Scanner::identifier() {
    while (isAlphaNumeric(peek())) advance();

    static const std::unordered_map<std::string, TokenType> keywordMap = {
        {"and", AND}, {"class", CLASS}, {"else", ELSE}, {"false", FALSE},
        {"func", FUNC}, {"for", FOR}, {"if", IF}, {"none", NONE}, {"or", OR},
        {"print", PRINT}, {"return", RETURN}, {"super", SUPER}, {"this", THIS},
        {"true", TRUE}, {"var", VAR}, {"while", WHILE}
    };

    std::string str = source.substr(start, current - start);
    auto it = keywordMap.find(str);
    if (it != keywordMap.end()) {
        addToken(it->second);
        return;
    }

    addToken(IDENTIFIER);
} 

void Scanner::number() {
    while (isDigit(peek())) advance();

    if (peek() == '.' && isDigit(peek(1))) {
        advance();
        while (isDigit(peek())) advance();

        std::string val = source.substr(start, current - start);
        Object value = Object(std::stod(val));
        addToken(TokenType::NUMBER, value);
        return;
    }

    std::string val = source.substr(start, current - start);
    Object value = Object(std::stoi(val));
    addToken(TokenType::NUMBER, value);
    return;
}

void Scanner::string() { // TODO: make normal strings vs multi line strings
    while (peek() != '"' && !isAtEnd()) {
        if (peek() == '\n') line++;
        advance();
    }
    if (isAtEnd()) {
        error(line, "Unterminated string");
        return;
    }
    advance();

    Object value = Object(source.substr(start + 1, current - start - 2));
    addToken(TokenType::STRING, value);
}

void Scanner::scanToken() {
    char c = advance();
    switch (c) {
        case '(': addToken(TokenType::LEFT_PAREN); break;
        case ')': addToken(TokenType::RIGHT_PAREN); break;
        case '{': addToken(TokenType::LEFT_BRACE); break;
        case '}': addToken(TokenType::RIGHT_BRACE); break;
        case '[': addToken(TokenType::LEFT_BRACKET); break;
        case ']': addToken(TokenType::RIGHT_BRACKET); break;
        case ',': addToken(TokenType::COMMA); break;
        case ';': addToken(TokenType::SEMICOLON); break;
        case ':': addToken(TokenType::COLON); break;
        case '@': addToken(TokenType::AT_SIGN); break;
        case '?': addToken(TokenType::QUESTION_MARK); break;

        case '!':
            addToken(match('=') ? TokenType::BANG_EQUAL : TokenType::BANG);
            break;
        case '=':
            addToken(match('=') ? TokenType::EQUAL_EQUAL : TokenType::EQUAL);
            break;
        case '<':
            addToken(match('=') ? TokenType::LESS_EQUAL : TokenType::LESS);
            break;
        case '>':
            addToken(match('=') ? TokenType::GREATER_EQUAL : TokenType::GREATER);
            break;
        case '%':
            addToken(match('=') ? TokenType::PERCENT_EQUAL : TokenType::PERCENT);
            break;
        case '^':
            addToken(match('=') ? TokenType::CARET_EQUAL : TokenType::CARET);
            break;
        case '~':
            addToken(match('=') ? TokenType::TILDE_EQUAL : TokenType::TILDE);
            break;

        case '*':
            addToken(match('*') ? TokenType::STAR_STAR :
                     match('=') ? TokenType::STAR_EQUAL :
                                  TokenType::STAR);
            break;

        case '&':
            addToken(match('&') ? TokenType::AMPERSAND_AMPERSAND :
                     match('=') ? TokenType::AMPERSAND_EQUAL :
                                  TokenType::AMPERSAND);
            break;

        case '|':
            addToken(match('|') ? TokenType::BAR_BAR :
                     match('=') ? TokenType::BAR_EQUAL :
                                  TokenType::BAR);
            break;

        case '+':
            addToken(match('+') ? TokenType::PLUS_PLUS :
                     match('=') ? TokenType::PLUS_EQUAL :
                                  TokenType::PLUS);
            break;
            
        case '-':
            addToken(match('-') ? TokenType::MINUS_MINUS :
                     match('>') ? TokenType::ARROW :
                     match('=') ? TokenType::MINUS_EQUAL :
                                  TokenType::MINUS);
            break;

        case '/':
            if (match('/'))
                while (peek() != '\n' && !isAtEnd())
                    advance();
            else
                addToken(match('=') ? TokenType::SLASH_EQUAL : TokenType::SLASH);
            break;

        case '.': 
            if (isDigit(peek())) {
                current--;
                number();
            } else addToken(TokenType::DOT);
            break;

        case '"': string(); break;

        case ' ':
        case '\r':
        case '\t':
            break;

        case '\n':
            line++;
            break;

        default:
            if (isDigit(c)) number();
            else if (isAlpha(c)) identifier();
            else error(line, "Unexpected character.");
            break;
    }
}

std::vector<Token> Scanner::scanTokens() {
    while (!isAtEnd()) {
        start = current;
        scanToken();
    }
    tokens.push_back({TokenType::END_OF_FILE, line, "", Object()});
    return tokens;
}

File: src/scanner.py
from src.error import error
from src.token import TokenType, Token


class Scanner:
    def __init__(self, source: str):
        self.source: str = source
        self.tokens: list[Token] = []

        self.start: int = 0
        self.current: int = 0
        self.line: int = 1

    def isDigit(self, char: str) -> bool:
        return all("0" <= c <= "9" for c in char)

    def isAlpha(self, char: str) -> bool:
        return all("a" <= c <= "z" or "A" <= c <= "Z" or c == '_' for c in char)

    def isAlphaNumeric(self, char: str) -> bool:
        return self.isDigit(char) or self.isAlpha(char)

    def isAtEnd(self) -> bool:
        return self.current >= len(self.source)

    def advance(self, n: int = 1) -> str:
        self.current += n
        return self.source[self.current - n: self.current]

    def peek(self, n: int = 1) -> str:
        if (self.current + n > len(self.source)):
            return '\0'
        return self.source[self.current: self.current + n]

    def match(self, expected: str) -> bool:
        if self.isAtEnd():
            return False
        if not self.source[self.current:].startswith(expected):
            return False
        self.current += len(expected)
        return True

    def addToken(self, tokenType: TokenType, value: any = None) -> None:
        self.tokens.append(Token(tokenType, self.source[self.start:self.current], value, self.line))

    def blockComment(self):
        nesting = 1
        while nesting > 0:
            if self.match("/*"):
                nesting += 1

            if self.match("*/"):
                nesting -= 1
                if nesting == 0:
                    break

            self.advance()

    def identifier(self) -> None:
        while self.isAlphaNumeric(self.peek()):
            self.advance()

        keywords = {
            "class": TokenType.CLASS,
            "else": TokenType.ELSE,
            "enum": TokenType.ENUM,
            "false": TokenType.FALSE,
            "for": TokenType.FOR,
            "func": TokenType.FUNC,
            "if": TokenType.IF,
            "interface": TokenType.INTERFACE,
            "none": TokenType.NONE,
            "print": TokenType.PRINT,
            "private": TokenType.PRIVATE,
            "protected": TokenType.PROTECTED,
            "public": TokenType.PUBLIC,
            "return": TokenType.RETURN,
            "super": TokenType.SUPER,
            "this": TokenType.THIS,
            "true": TokenType.TRUE,
            "var": TokenType.VAR,
            "while": TokenType.WHILE
        }

        lexme = self.source[self.start:self.current]
        self.addToken(keywords.get(lexme, TokenType.IDENTIFIER))

    def number(self) -> None:
        while self.isDigit(self.peek()):
            self.advance()

        if self.peek() == '.' and (self.isDigit(self.peek(2)[1])):
            self.advance()
            while self.isDigit(self.peek()):
                self.advance()

        self.addToken(TokenType.NUMBER, float(self.source[self.start:self.current]))

    def string(self) -> None:
        if self.match("\"\""):
            while self.peek(3) != "\"\"\"" and not self.isAtEnd():
                if self.peek() == "\n":
                    self.line += 1
                self.advance()

            if self.isAtEnd():
                error(self.line, "Unterminated string.")
                return

            self.advance(3)
            self.addToken(TokenType.STRING, self.source[self.start + 3:self.current - 3])
            return

        while self.peek() != "\"" and not (self.isAtEnd() or self.peek() == '\n'):
            self.advance()

        if self.isAtEnd() or self.peek() == '\n':
            error(self.line, "Unterminated string.")
            return

        self.advance()
        self.addToken(TokenType.STRING, self.source[self.start + 1:self.current - 1])


    def scanToken(self) -> None:
        c: str = self.advance()
        if c == "(":   self.addToken(TokenType.LEFT_PAREN)
        elif c == ")": self.addToken(TokenType.RIGHT_PAREN)
        elif c == "{": self.addToken(TokenType.LEFT_BRACE)
        elif c == "}": self.addToken(TokenType.RIGHT_BRACE)
        elif c == "[": self.addToken(TokenType.LEFT_BRACKET)
        elif c == "]": self.addToken(TokenType.RIGHT_BRACKET)
        elif c == ',': self.addToken(TokenType.COMMA)
        elif c == '@': self.addToken(TokenType.AT_SIGN)
        elif c == ':': self.addToken(TokenType.COLON)
        elif c == ';': self.addToken(TokenType.SEMICOLON)
        elif c == '?': self.addToken(TokenType.QUESTION)

        elif c == "=":
            self.addToken(TokenType.EQUAL_EQUAL if self.match("=") else TokenType.EQUAL)
        elif c == "!":
            self.addToken(TokenType.BANG_EQUAL if self.match("=") else TokenType.BANG)
        elif c == "~":
            self.addToken(TokenType.TILDE_TILDE if self.match("~") else TokenType.TILDE)
        elif c == "^":
            self.addToken(TokenType.CARET_EQUAL if self.match("=") else TokenType.CARET)
        elif c == "%":
            self.addToken(TokenType.PERCENT_EQUAL if self.match("=") else TokenType.PERCENT)

        elif c == "/":
            if self.match("/"):
                while self.peek() != '\n' and not self.isAtEnd():
                    self.advance()
            elif self.match("*"):
                self.blockComment()
            elif self.match("="):
                self.addToken(TokenType.SLASH_EQUAL)
            else:
                self.addToken(TokenType.SLASH)

        elif c == "*":
            if self.match("*"):
                self.addToken(TokenType.STAR_STAR)
            elif self.match("="):
                self.addToken(TokenType.STAR_EQUAL)
            else:
                self.addToken(TokenType.STAR)

        elif c == "+":
            if self.match("+"):
                self.addToken(TokenType.PLUS_PLUS)
            elif self.match("="):
                self.addToken(TokenType.PLUS_EQUAL)
            else:
                self.addToken(TokenType.PLUS)

        elif c == "|":
            if self.match("|"):
                self.addToken(TokenType.BAR_BAR)
            elif self.match("="):
                self.addToken(TokenType.BAR_EQUAL)
            else:
                self.addToken(TokenType.BAR)

        elif c == "&":
            if self.match("&"):
                self.addToken(TokenType.AMPERSAND_AMPERSAND)
            elif self.match("="):
                self.addToken(TokenType.AMPERSAND_EQUAL)
            else:
                self.addToken(TokenType.AMPERSAND)

        elif c == "-":
            if self.match("-"):
                self.addToken(TokenType.MINUS_MINUS)
            elif self.match(">"):
                self.addToken(TokenType.ARROW)
            elif self.match("="):
                self.addToken(TokenType.MINUS_EQUAL)
            else:
                self.addToken(TokenType.MINUS)

        elif c == "<":
            if self.match("<="):
                self.addToken(TokenType.LESS_LESS_EQUAL)
            elif self.match("<"):
                self.addToken(TokenType.LESS_LESS)
            elif self.match("="):
                self.addToken(TokenType.LESS_EQUAL)
            else:
                self.addToken(TokenType.LESS)

        elif c == ">":
            if self.match(">="):
                self.addToken(TokenType.GREATER_GREATER_EQUAL)
            elif self.match(">"):
                self.addToken(TokenType.GREATER_GREATER)
            elif self.match("="):
                self.addToken(TokenType.GREATER_EQUAL)
            else:
                self.addToken(TokenType.GREATER)

        elif self.isAlpha(c):
            self.identifier()

        elif c == "\"":
            self.string()

        elif c == '.':
            if self.isDigit(self.peek()):
                self.number()
            else:
                self.addToken(TokenType.DOT)

        elif self.isDigit(c):
            self.number()

        elif c == '\n':
            self.line += 1

        elif c.isspace():
            pass

        else:
            error(self.line, f"Unexpected character: {c}.")

    def scanTokens(self) -> list[Token]:
        while not self.isAtEnd():
            self.start = self.current
            self.scanToken()

        self.tokens.append(Token(TokenType.EOF, "", None, self.line))
        return self.tokens
